### DISCUSSION TITLE
How to `ref.refresh` while discarding previous state?

### DISCUSSION THREAD

**Discussion by fujidaiti** [rrousselGit/riverpod#3691] [Q&A]
When refreshing a`FutureProvider` using `ref.refresh`, it keeps the previous state while the asynchronous process is running. For example, if the previous state is an `AsyncData`, `ref.refresh` changes the provider's state to `AsyncData(isLoading: true)`, not to `AsyncLoading`. This behavior makes sense for cases like pull-to-refresh actions (showing old values while loading).

However, there are some cases where this behavior is not ideal. For instance, when recovering from an error state such as a network connection failure, we might want to restart the provider from the `AsyncLoading` state rather than keeping the old error state. For `Notifier`s, we can manually achieve this as follows:

```dart
Future<void> refresh() async {
  state = const AsyncLoading();
  // Fetch data again
  ...
}
```

This approach works well for `Notifier`s, but it's a bit tedious and doesn't work for `FutureProvider`s. Is there a more efficient way to implement this restarting behavior?

Thank you.

**Comment by johnnyggalt** [comment]
> For Notifiers, we can manually achieve this as follows
Actually, I don't think that's the case. I arrived at trying exactly that before giving up and ending up in this thread.

The problem is that the `state` property invokes `_element.state = newState;`, and that ends up invoking `AsyncTransition<T>.asyncTransition`, which insists on using `copyWithPrevious` if there is any previous state. There is no way to wipe the slate clean with respect to state.

My scenario is that when the user adds a new item, I'd prefer to just have the UI reset back to what happens when it first loads.

**Comment by rrousselGit** [comment]
You generally shouldn't try to clear the "data" of an AsyncValue.

Instead, prefer having your UI check `asyncValue.isRefreshing`
